### PR TITLE
Add SICP JS solutions

### PR DESCRIPTION
**Pull request by arsalan0c**
This PR adds two solutions for the n-queens puzzle using the Source 3 non-deterministic language. Each of them makes use of the generate-and-test paradigm.
<br />
The first (`queens_slow`) uses non-determinism for generating both the row and column for each position. It does so in an optimized manner, testing the corresponding condition as soon as the choice is generated, thereby reducing the search space. However, this is not enough to yield a quick solution when N = 8. This solution also gives repeated solutions (i.e different permutations that have all the same positions) upon backtracking.
<br />
The second (`queens_fast`) uses non-determinism in picking only the row and not the column of each position, with the columns being fixed. This further reduces the search space, yielding a quick solution when N = 8.

Example:
image

### CODE REVIEW

File: src/examples/source-3-non-det/queens.js [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+// const result_queens_slow = queens_slow(N, N);
+// pretty_print(result_queens_slow, N);
+// generate more solutions by entering 'try again' in the REPL
+
+function queens_slow(board_size, num_queens) {
+  require(num_queens <= board_size);
+  const possible_positions = enum_list(1, board_size);
+
+  const result = accumulate(
```
Comment (anubh-v) [inline src/examples/source-3-non-det/queens.js:33]: Is there a reason to use accumulate, other than the fact that we want to perform the process `num_queens` times?
If not, can \ should we use `for_each` ?

File: src/examples/source-3-non-det/liars.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+      : is_null(tail(items))
+      ? true
+      : is_null(member(head(items), tail(items)))
+          ? distinct(tail(items))
+          : false;
+}
+
+const kitty = amb(1, 2, 3, 4, 5);
+const mary = amb(1, 2, 3, 4, 5);
+require(amb(kitty === 2, mary === 4));
```
Comment (anubh-v) [inline src/examples/source-3-non-det/liars.js:19]: To reduce the search space, I have done the `require` checks immediately once the information needed for the check is available.

File: src/examples/source-3-non-det/queens.js (hunk 1 above)
Comment (arsalan0c) [reply src/examples/source-3-non-det/queens.js:33]: Yes, accumulate gives us the positions that we have already, with the variable `so_far`. It is used in checking whether potential new rows and columns can be added.

File: src/examples/source-3-non-det/liars.js (hunk 2 above)
Comment (arsalan0c) [reply src/examples/source-3-non-det/liars.js:19]: Great.

**Comment by arsalan0c** [comment]
See https://github.com/source-academy/source-programs/pull/16

**Pull request closed without merging**